tests/vm_generator.go

```diff
@@ -0,0 +1,95 @@
+package tests
+
+import (
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+
+	v1 "kubevirt.io/api/core/v1"
+)
+
+func randName(name string) string {
+	return name + "-" + rand.String(5)
+}
+
+func getBaseVMISpec() *v1.VirtualMachineInstanceSpec {
+	return &v1.VirtualMachineInstanceSpec{
+		Domain: v1.DomainSpec{
+			Resources: v1.ResourceRequirements{
+				Requests: k8sv1.ResourceList{
+					k8sv1.ResourceMemory: resource.MustParse("64M"),
+				},
+			},
+		},
+	}
+}
+
+func addContainerDisk(spec *v1.VirtualMachineInstanceSpec, image string) *v1.VirtualMachineInstanceSpec {
+	disk := &v1.Disk{
+		Name: "containerdisk",
+		DiskDevice: v1.DiskDevice{
+			Disk: &v1.DiskTarget{
+				Bus: "virtio",
+			},
+		},
+	}
+	spec.Domain.Devices.Disks = append(spec.Domain.Devices.Disks, *disk)
+	volume := &v1.Volume{
+		Name: "containerdisk",
+		VolumeSource: v1.VolumeSource{
+			ContainerDisk: &v1.ContainerDiskSource{
+				Image: image,
+			},
+		},
+	}
+	spec.Volumes = append(spec.Volumes, *volume)
+	return spec
+}
+
+func getBaseVMI(name string) *v1.VirtualMachineInstance {
+	baseVMISpec := getBaseVMISpec()
+
+	return &v1.VirtualMachineInstance{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: v1.GroupVersion.String(),
+			Kind:       "VirtualMachineInstance",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   name,
+			Labels: map[string]string{"special": name},
+		},
+		Spec: *baseVMISpec,
+	}
+}
+
+func addNoCloudDiskWitUserData(spec *v1.VirtualMachineInstanceSpec, data string) *v1.VirtualMachineInstanceSpec {
+	spec.Domain.Devices.Disks = append(spec.Domain.Devices.Disks, v1.Disk{
+		Name: "cloudinitdisk",
+		DiskDevice: v1.DiskDevice{
+			Disk: &v1.DiskTarget{
+				Bus: "virtio",
+			},
+		},
+	})
+
+	spec.Volumes = append(spec.Volumes, v1.Volume{
+		Name: "cloudinitdisk",
+		VolumeSource: v1.VolumeSource{
+			CloudInitNoCloud: &v1.CloudInitNoCloudSource{
+				UserData: data,
+			},
+		},
+	})
+	return spec
+}
+
+func CreateVmiObject(name string, namespace string, interfaces []v1.Interface, networks []v1.Network) *v1.VirtualMachineInstance {
+	vmi := getBaseVMI(randName(name))
+	vmi.Namespace = namespace
+	addContainerDisk(&vmi.Spec, "kubevirt/cirros-container-disk-demo:latest")
+	addNoCloudDiskWitUserData(&vmi.Spec, "#!/bin/sh\n\necho 'printed from cloud-init userdata'\n")
```

Comment:
Lets divide into functions please if possible
```
func addNoCloudDisk(spec *v1.VirtualMachineInstanceSpec) *v1.VirtualMachineInstanceSpec {
	return addNoCloudDiskWitUserData(spec, "#!/bin/sh\n\necho 'printed from cloud-init userdata'\n")
}
```

Reply:
why?

Reply:
it is nicer to have functions that hide the details and you just need to read the name when you read top down

Reply:
I find it more readable to see we are adding here a NoCloudDisk with UserData. I don't see the value in hiding this detail.

Reply:
the more important thing here is the why
without it it the cirros is taking more time to load, something about the ipv4 stack
i think otherwise but won't insist